### PR TITLE
Prevent memory dumps on windows.

### PR DESCRIPTION
**Pull request by rockihack**
## Description

Use a discretionary access control list to deny PROCESS_QUERY_INFORMATION and PROCESS_VM_READ access rights without admin privileges.

## Motivation and Context

Prevent memory dumps on windows

## How Has This Been Tested?

Windows 7 and Process Explorer
**Needs more testing**

## Types of changes

- ✅ New feature (non-breaking change which adds functionality)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**

### CODE REVIEW

**Comment by phoerious** [comment]
Is that necessary? Isn't the call to `Tools::disableCoreDumps()` in main.cpp:44 enough? Or is there some more magic involved on Windows?

**Comment by rockihack** [comment]
`Tools::disableCoreDumps()` doesn't do anything on windows yet...

**Comment by phoerious** [comment]
It does not? Hmm. Interesting.
Too bad, I don't really understand a line of this low-level Windows API-specific C code there. :-\
Are those gotos really necessary?

**Comment by rockihack** [comment]
Gotos are never really necessary, but I use them to minimize nesting.
In kernel development they are common for cleanup purposes.

**Comment by phoerious** [comment]
Well, I would prefer returns here. I don't see the advantage of using goto in this case.
Also Travis is having some remarks for you.

**Comment by rockihack** [comment]
Feel free to replace the gotos with returns and duplicate the cleanup code six times.

**Comment by phoerious** [comment]
What not put the cleanup code after the function call? Then you only need it once. No need to duplicate anything.

BTW you should rebase the branch. We merged some big stuff recently.

**Comment by rockihack** [comment]
If you think it is necessary to remove the gotos then go ahead. I think it produces cleaner code.

**Comment by phoerious** [comment]
It breaks with the coding style of the rest of the application. C++ is not really the language you want to use goto in. I realize that part is more low-level C code, but we should still maintain some style consistency.

**Comment by rockihack** [comment]
Normally you use `RAII` concept in c++ if you want it to be consistent.

**Comment by phoerious** [comment]
Which is a little difficult with state machine code like that. Still, I'm not too convinced about goto.

**Comment by droidmonkey** [comment]
My problem with this PR is that it is not very well explained (in code or in the PR). Where did you get this process from? Is this endorsed anywhere? I did a cursory search for disabling core dumps on windows and came up empty handed, certainly nothing popped up like what you present here.

I need to understand more before this will be incorporated.

**Comment by rockihack** [comment]
There is pretty much nothing what you can find about disabling memory dumps on windows. I figured it out myself. The reason I opened the PR is to get more people involved to test it on different machines and operating systems (windows 8, windows 10).

You can test it yourself with ProcessExplorer.
1) Open ProcessExplorer
2) Rightclick on KeePassX.exe
3) Choose Create Dump -> Mini / Full Dump
4) An "Access Denied" message box pops up

With this in mind you first need to know how memory dumping works on windows from usermode.
The first step is to open a process handle (OpenProcess) with atleast PROCESS_QUERY_INFORMATION and PROCESS_VM_READ access rights,
because MiniDumpWriteDump requires you to do so.
see: https://msdn.microsoft.com/en-us/library/windows/desktop/ms680360%28v=vs.85%29.aspx

If you deny to open the process with those rights, you can't dump memory anymore.
How to do it on windows? With a discretionary access control list.
This works as long as you do not have admin privileges, since then you are able to grant yourself the SE_TAKE_OWNERSHIP_NAME or SE_DEBUG_NAME privilege and circumvent the DACL (you cannot protect something from root).

So basically this PR uses a DACL to grant the creator of the process minimal access rights and deny everything else. Effectively this prevents memory dumps on windows, but might have some other side affects I'm not aware of.

**Comment by rockihack** [comment]
I explained my intentions above to prevent accessing memory from other applications.
Memory dumps can also occur through windows error reporting when the application crashes.
This feature is not enabled by default according to [2].

Especially **WerAddExcludedApplication** might be interesting.
[1] https://msdn.microsoft.com/en-us/library/windows/desktop/bb513617(v=vs.85).aspx

[2] **Collecting User-Mode Dumps:** https://msdn.microsoft.com/en-us/library/windows/desktop/bb787181(v=vs.85).aspx

[3] **Using WER:** https://msdn.microsoft.com/en-us/library/windows/desktop/bb513616.aspx

**Comment by phoerious** [comment]
Thanks for the explanation, that makes things clearer.
I think, core dumping (and therefore data leakage to the hard disk) due to crashes is the bigger issue and also the main reason why it's disable on Linux.

**Comment by rockihack** [comment]
It seems my pull request is already working as intended.

1) First enable error reporting for KeePassX.exe explicitly.
https://msdn.microsoft.com/en-us/library/windows/desktop/bb787181(v=vs.85).aspx

2) Patch main.cpp
```
`#ifdef QT_NO_DEBUG
    Tools::disableCoreDumps();

    // Segment fault
    int* foo = nullptr;
    *foo = 0;
#endif
```

3) Start KeePassX (without DACL)
=> It crashes right away and produces a mini dump in `%LOCALAPPDATA%\CrashDumps`.

4) Start KeePassX (with DACL)
=> No crash dump is created.

**Comment by rockihack** [comment]
@phoerious Using a DACL also prevents memory access from other applications.
Of course crash dumps written to the harddrive are a security issue, but on windows it's even possible to access other process' memory. Thats insane.

**Comment by phoerious** [comment]
That I did not know. That actually IS insane.

**Comment by droidmonkey** [comment]
OK great explanation and excellent work. This should probably stand-alone as an example of preventing crash dumps in windows (ie on Stack Overflow).

File: src/core/Tools.cpp [hunk 1]
```diff
@@ -240,4 +246,110 @@ void setupSearchPaths()
 #endif
 }
 
+//
+// Prevent memory dumps without admin privileges.
+// MiniDumpWriteDump function requires PROCESS_QUERY_INFORMATION and PROCESS_VM_READ
+// see: https://msdn.microsoft.com/en-us/library/windows/desktop/ms680360%28v=vs.85%29.aspx
+//
```
Comment (droidmonkey) [inline src/core/Tools.cpp]: Add the following to this comment:

>If you deny to open the process PROCESS_QUERY_INFORMATION and PROCESS_VM_READ access rights, you can't dump memory anymore. We do this using discretionary access control list (DACL) on Windows.
>
>This works as long as you do not have admin privileges, since then you are able to grant yourself the SE_TAKE_OWNERSHIP_NAME or SE_DEBUG_NAME privilege and circumvent the DACL.
>
>This function grants the process minimal access rights and denies everything else. Effectively this prevents memory dumps on windows and disallows other processes from accessing our memory.

Comment (rockihack) [inline src/core/Tools.cpp]: I will add a more detailed comment tomorrow.
Please verify that it's working with windows 10 too.

Comment (phoerious) [inline src/core/Tools.cpp]: I just tested it on Windows 10 and it doesn't seem to work. I opened the task manager, right-clicked KeePassXC and chose "Create dump file". After a short while it said "The file has been successfully created."

**Comment by droidmonkey** [comment]
It is true that you can create a dump file from the task manager in windows 10, but the forced crash does NOT create a dump file in %localappdata%\CrashDumps

**Comment by rockihack** [comment]
On windows 10 `taskmgr.exe` is running with `SeDebugPrivilege` and therefore is able to circumvent the DACL. You can verify this by using ProcessExplorer.
https://technet.microsoft.com/en-us/sysinternals/processexplorer.aspx

Maybe we should use **WerAddExcludedApplication** just in case.

**Comment by droidmonkey** [comment]
We can do that with the installer

**Comment by TheZ3ro** [comment]
But that will require to run the installer with admin privilege. 
Can someone on windows tests how KeePass behave with ProcessExplorer? I think having the DACL is enought (if an expert users want, he can edit the registry himself)

**Comment by rockihack** [comment]
I added a more detailed comment.

Personally I don't use keepassxc but maintain my own fork.
It would be great if you create a list of security improvements compared to keepassx.

**Comment by droidmonkey** [comment]
The change logs for KeePassXC clearly indicate what was done in each release. Recommend looking there. If you want to consolidate them, please open an issue and we will look to move them into the wiki.

[ChangeLog](https://github.com/keepassxreboot/keepassxc/blob/develop/CHANGELOG)

**Comment by phoerious** [comment]
Any specific reason for using a custom fork and not the mainline?

**Comment by droidmonkey** [comment]
@rockihack does this answer the mail for you: https://github.com/keepassxreboot/keepassxc/wiki/KeePassXC-Changelog

**Comment by rockihack** [comment]
I mean a fork of keepassx. I prefer stable and secure software over fancy usability stuff.
And I think that it's just a matter of time that keepassxc becomes bloatware with too many features like the original keepass. As an example I don't want any networking code in my password manager like the favicon downloader or even worse KeePassHTTP.

**Comment by phoerious** [comment]
> I prefer stable and secure software over fancy usability stuff.

I wouldn't exactly call KeePassX 2.0 "stable". It has quite some annoying bugs, which is understandable considering that is was basically a rewrite. KeePassX 0.4 was stable.

> And I think that it's just a matter of time that keepassxc becomes bloatware with too many features like the original keepass.

Harsh words. 😞
